services/camera/libcameraservice/CameraService.cpp

```diff
@@ -387,12 +385,14 @@ static void torch_mode_status_change(
 // This is ugly and only safe if we never re-create the CameraService, but
 // should be ok for now.
 static CameraService *gCameraService;
+static android::sp<TrustAgentRegistry> trust_agent_registry;
 
 CameraService::CameraService() :
         mEventLog(DEFAULT_EVENT_LOG_LENGTH),
         mNumberOfCameras(0), mNumberOfNormalCameras(0),
         mSoundRef(0), mModule(nullptr) {
     ALOGI("CameraService started (pid=%d)", getpid());
+    trust_agent_registry = android::sp<TrustAgentRegistry>(new TrustAgentRegistry("/dev/socket/camera_service/camera_service_to_trust"));
```

Comment:
If `CameraService` really happens to be instantiated twice, then `TrustAgentRegistry` will still be instantiated twice, just that it'll be in place of the old one. Maybe move the instantiation to the declaration above instead (should be safe since C++11 guarantee no race). Alternative is to implement `TrustAgentRegistry::instance()`, but that would be a bit too much diff...

Reply:
It might occur to you after a deeper inspection with lsof that mediaserver, cameraserver, camera_service & minimedia on the Xperia X with the *status quo* implementation all attempt to create the socket, hence have the TrustAgentRegistry instantiated *without* CameraService being involved. Nothing is supposed to spawn another CameraService other than either camera_service or minimedia (depending on Android version).

Reply:
Ah, I misread the diff and thought the line you removed were in the `CameraService`'s class definition. My bad. 😅

That said, this is still a little bit tricky. Moving this into `CameraService`'s class definition is probably the best, but it isn't really feasible since it's defined in `CameraService.h`. Having `TrustAgentRegistry::instance()` that instantiate on first call should work for this purpose, but it's still a diff. Another way is to check if it's already created before creating a new one, but it's unclean...

> [...] Nothing is supposed to spawn another CameraService other than either camera_service or minimedia [...]

You know, on Android `cameraserver` _is_ supposed to host `CameraService`. We didn't comment out this instantiation in `cameraserver` [1] so with 3rd commit alone it won't be enough (And in fact, we didn't comment out one in `mederserver` either [2]).

So we need the first 2 commits because it's not what _supposed_ to happen that matters, it's what _actually_ happens.

On basis of minimal diff I _almost_ tempted to accept only first 2 commits and not the third. However it might occur that we need `mediaserver` for some other things in the future. So it probably doesn't hurt to also have the 3rd one.

[1] https://github.com/Halium/android_frameworks_av/blob/halium-7.1/camera/cameraserver/main_cameraserver.cpp
[2] https://github.com/Halium/android_frameworks_av/blob/halium-7.1/media/mediaserver/main_mediaserver.cpp#L48

Reply:
I don't have my Xperia X without those patches with me right now, but yeah, I was able to identify those 4 components that all link against libcameraservice and trip over each other during socket creation.

I don't know, maybe stuffing the `TrustAgentRegistry` in a singleton might be a good thing but then again, neither is `CameraService` since it makes itself the global pointer in its ctor, and I doubt any consumer of libcameraservice instantiates more than one `CameraService`.

Reply:
Hmm right, forgot about `minimedia`. Then the 3rd commit is likely also needed.

I won't nitpick over this stuff anymore then. Let's get this in!